Tweak to support Microsoft-style archives

iximeow opened the pull request:
I happened to run into this looking at `msvcrt.lib` and friends, the name table consists of a bunch of null-terminated strings, rather than /-terminated as the code currently expects.

It now *appears* to work on these files, but I haven't tested too heavily. I'm also not sure how legit it is by MS licensing to keep a copy of one of their `.lib` apart from VS to test against.

mdsteele commented:
Thanks!  Could you also add a new unit test for this?  (e.g. similar to `read_gnu_archive_with_long_filenames()`, but with null-terminated name table strings)

FWIW, I dug a bit and found this: https://msdn.microsoft.com/en-us/library/windows/desktop/ms680547(v=vs.85).aspx#longnames_member, which seems to confirm that the name table uses null-terminated strings for `.lib` files: "The strings are null-terminated. Each string begins immediately after the null byte in the previous string."  (Although, that doc is apparently a bit inconsistent in some places -- in that subsection it says the name table uses the key "\\\\", but in a previous subsection it says it uses "//", as GNU ar does.  Sounds like maybe "//" is correct, since that's what this crate assumes, and it seems to work?)

iximeow commented:
Gee where'd those two weeks go?

I added a test for MS-style archives, which reminded me that there was a little bit more to fix: Microsoft tools emit strings of spaces for `uid` and `gid` fields - not even `0       `. This is backed up by that MSDN page you'd found earlier: `An ASCII decimal representation of the user ID. This field does not  contain a meaningful value on Windows platforms because Microsoft tools  emit all blanks.`

The test case I added includes this, and my solution was to just pretend empty string means 0, so MS `uid`/`gid` are interpreted as `0`'s. I don't ever actually extract anything, so permissions don't matter so much, but I haven't thought through the implications of this at all.

If this sounds reasonable to you, it works for me!

mdsteele commented:
Yep, defaulting to zero sounds totally reasonable.  Even on *nix systems, archive files I've seen often just have zero for the UID/GID.

mdsteele commented:
Thanks for the changes!
